File: python/oauth_tls.py
import os
import pulsar

if __name__ == '__main__':
    service_url = os.environ['PULSAR_SERVICE_URL']
    topic = os.environ['PULSAR_TOPIC']
    issuer_url = os.environ['PULSAR_ISSUER_URL']
    private_key = os.environ['PULSAR_PRIVATE_KEY']
    audience = os.environ['PULSAR_AUDIENCE']

    auth = pulsar.AuthenticationOauth2(f'''{{
        "issuer_url": "{issuer_url}",
        "private_key": "{private_key}",
        "audience": "{audience}"
    }}''')

    client = pulsar.Client(service_url,
                           authentication=auth,
                           tls_allow_insecure_connection=False,
                           tls_validate_hostname=True)
    producer = client.create_producer(topic)
    msg_id = producer.send('hello'.encode())
    print(f'Sent to {msg_id}')
    producer.close()
    client.close()

File: node/oauth_tls.js
const Pulsar = require('pulsar-client');

(async () => {
    Pulsar.Client.setLogHandler((level, file, line, message) => {
        console.log('[%s][%s:%d] %s', Pulsar.LogLevel.toString(level), file, line, message);
    });

    const auth = new Pulsar.AuthenticationOauth2({
        issuer_url: process.env.PULSAR_ISSUER_URL,
        private_key: process.env.PULSAR_PRIVATE_KEY,
        audience: process.env.PULSAR_AUDIENCE,
    })

    const client = new Pulsar.Client({
        serviceUrl: process.env.PULSAR_SERVICE_URL,
        authentication: auth,
    });

    const producer = await client.createProducer({
        topic: process.env.PULSAR_TOPIC,
    });

    const msg_id = await producer.send({
        data: Buffer.from(`hello`),
    });
    console.log(msg_id)

    await producer.close();
    await client.close();
})();
